### PR TITLE
AP_Mount: correct double-mapping of port to channel number

### PR DESCRIPTION
**Pull request by peterbarker**
Tested on a PixyU connected to SITL.

Note that most of the nullptr checks are can't-happens due to checks elsewhere.

Note also that we can save some bytes in the send-message methods in future by using `link->set_message(type, pkt)`.   Not done here to make backport easier.

This work sponsored by Harris Aerial.

### CODE REVIEW

File: libraries/AP_Mount/AP_Mount_Gremsy.cpp [hunk 1]
```diff
@@ -304,6 +322,11 @@ void AP_Mount_Gremsy::send_gimbal_device_set_attitude(float roll_rad, float pitc
         return;
     }
 
+    if (_link == nullptr) {
+        return;
+    }
+    const mavlink_channel_t _chan = _link->get_chan();
```
Comment (rmackay9) [inline libraries/AP_Mount/AP_Mount_Gremsy.cpp:328]: Let's replace "_chan" with "chan" 'cuz normally the underscore is used to signify that the variable is an internal variable.  I don't mind a little more code churn.

File: libraries/AP_Mount/AP_Mount_Gremsy.cpp [hunk 2]
```diff
@@ -152,17 +152,16 @@ void AP_Mount_Gremsy::find_gimbal()
     }
 
     // search for a mavlink enabled gimbal
-    if (!_found_gimbal) {
+    if (_link == nullptr) {
         // we expect that instance 0 has compid = MAV_COMP_ID_GIMBAL, instance 1 has compid = MAV_COMP_ID_GIMBAL2, etc
         uint8_t compid = (_instance == 0) ? MAV_COMP_ID_GIMBAL : MAV_COMP_ID_GIMBAL2 + (_instance - 1);
-        if (GCS_MAVLINK::find_by_mavtype_and_compid(MAV_TYPE_GIMBAL, compid, _sysid, _chan)) {
-            _compid = compid;
-            _found_gimbal = true;
```
Comment (rmackay9) [inline libraries/AP_Mount/AP_Mount_Gremsy.cpp:160]: we should be able to remove the _found_gimbal variable as well I think.

**Comment by rmackay9** [comment]
So the underlying issue is that we were incorrectly passing a channel number into the set_message_interval() method that actually takes a serial port number.  This means the set_message_interval would be run on the wrong serial port if there wasn't a 1-to-1 mapping between channel and serial port number.  So for example if the gimbal was connected to Serial5 which is normally channel 2 (I think).

An alternative solution would be to add a method to convert chan to serial port num (the reverse already exists).

File: libraries/AP_Mount/AP_Mount_Gremsy.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Mount/AP_Mount_Gremsy.cpp:328]: Done

File: libraries/AP_Mount/AP_Mount_Gremsy.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Mount/AP_Mount_Gremsy.cpp:160]: Already done :-)

**Comment by peterbarker** [comment]
> So the underlying issue is that we were incorrectly passing a channel number into the set_message_interval() method that actually takes a serial port number. This means the set_message_interval would be run on the wrong serial port if there wasn't a 1-to-1 mapping between channel and serial port number. So for example if the gimbal was connected to Serial5 which is normally channel 2 (I think).

channel 3 IIUC, assuming 0,1 and 2 channels correspond to serials 0, 1 and 2 which is our default setup.

**Review by rmackay9 — APPROVED** [review]
As long as it's been tested then I'm happy with this, thanks very much for finding and fixing this bug!

**Comment by Davidsastresas** [comment]
Good catch!

**Comment by rmackay9** [comment]
This is included in Copter-4.3.5-rc1

**Pull request merged**